### ISSUE TITLE
stage-1: no longer able to mount multi-device btrfs

### ISSUE THREAD

**Issue by Atemu** [NixOS/nixpkgs#265668]
### Describe the bug

Stage 1 stops after LUKS unlock and the "starting LVM and device mapper" with an undescript error. Launching an interactive shell revealed that, while the LUKS devices are opened and available, trying to mount any of my two btrfs from the mapped devices directly does not work; always failing to find the other device (2x2 setup).  

### Steps To Reproduce
1. Have one btrfs spanning two LUKS devices 
2. Try to boot with 85f1ba3e51676fa8cc604a3d863d729026a6b8eb and boot.debug1devices
3. `mount /dev/mapper/devicenamehere /somewhere` in the debug shell

### Additional context

Does not happen with 63678e9f3d3afecfeafa0acead6239cdb447574c.

How the hell did this pass CI?

### Notify maintainers

@aszlig @lheckemann @nikstur 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

**Comment by Atemu** [comment]
This reproduces in `nixosTests.systemd-initrd-btrfs-raid`. Bisecting...

**Comment by Atemu** [comment]
No it doesn't. I had modified the test to not use systemd initrd and thought I removed those changes but it only seems to affect stage-1.

**Comment by Atemu** [comment]
Using 

```patch
diff --git a/nixos/tests/systemd-initrd-btrfs-raid.nix b/nixos/tests/systemd-initrd-btrfs-raid.nix
index 9196033789cb..89ef7176b0d0 100644
--- a/nixos/tests/systemd-initrd-btrfs-raid.nix
+++ b/nixos/tests/systemd-initrd-btrfs-raid.nix
@@ -14,10 +14,10 @@ import ./make-test-python.nix ({ lib, pkgs, ... }: {
     boot.loader.efi.canTouchEfiVariables = true;
 
     environment.systemPackages = with pkgs; [ btrfs-progs ];
-    boot.initrd.systemd = {
-      enable = true;
-      emergencyAccess = true;
-    };
+    # boot.initrd.systemd = {
+    #   enable = true;
+    #   emergencyAccess = true;
+    # };
 
     specialisation.boot-btrfs-raid.configuration = {
       fileSystems = lib.mkVMOverride {
@@ -39,7 +39,7 @@ import ./make-test-python.nix ({ lib, pkgs, ... }: {
     machine.wait_for_unit("multi-user.target")
 
     # Ensure we have successfully booted from the RAID
-    assert "(initrd)" in machine.succeed("systemd-analyze")  # booted with systemd in stage 1
+    # assert "(initrd)" in machine.succeed("systemd-analyze")  # booted with systemd in stage 1
     assert "/dev/vdb on / type btrfs" in machine.succeed("mount")
     assert "hello" in machine.succeed("cat /test")
     assert "Total devices 2" in machine.succeed("btrfs filesystem show")
```

bisect revealed f126ff2c29d747d6e013690f27baa3ae0814349f as the cause. cc @trofi

**Comment by Atemu** [comment]
What's interesting though is that the systemd-initrd does not seem to be affected? I'll verify that on the real system now.

Still, I think we should probably revert the btrfs-progs bump until the actual cause is known.

**Comment by aszlig** [comment]
> Still, I think we should probably revert the btrfs-progs bump until the actual cause is known.

It probably is also a good idea to add a regression test for this specific issue.

**Comment by Atemu** [comment]
The systemd-initrd test works for that if you disable the systemd-initrd part (see patch above). I don't think there's a good way to make it generic though without https://github.com/NixOS/nixpkgs/pull/176557/. Should I just copy paste?

**Comment by trofi** [comment]
Yeah, the boot fails to format devices as they are not ready yet:

```
machine # [    4.095032] Btrfs loaded, crc32c=crc32c-intel, zoned=no, fsverity=no
machine # ERROR: device scan failed on '/dev/vdb': No such file or directory
machine # ERROR: device scan failed on '/dev/vdc': No such file or directory
(finished: must succeed: mkfs.btrfs -d raid0 /dev/vdb /dev/vdc, in 10.31 seconds)
```

Upstream added an interesting `btrfs device ready` call to the CI as that also started failing for unavailable devices: https://github.com/kdave/btrfs-progs/commit/1faaa874a1263a4beb7d14b8ac2378526ac727eb

I'll try to bisect `btrfs-progs` itself.

**Comment by trofi** [comment]
Bisect points at seemingly harmless https://github.com/kdave/btrfs-progs/commit/fcd9142b67fe4e73d275b7d854bf58842bbe133a change but I think it has a smoking gun: in changed IOCTL definition:

```diff
--- a/kernel-shared/uapi/btrfs.h
+++ b/kernel-shared/uapi/btrfs.h
@@ -1171,7 +1171,7 @@ enum btrfs_err_code {
                                   struct btrfs_ioctl_vol_args)
 #define BTRFS_IOC_RESIZE _IOW(BTRFS_IOCTL_MAGIC, 3, \
                                   struct btrfs_ioctl_vol_args)
-#define BTRFS_IOC_SCAN_DEV _IOW(BTRFS_IOCTL_MAGIC, 4, \
+#define BTRFS_IOC_SCAN_DEV _IOW(BTRFS_IOCTL_MAGIC, 5, \
                                   struct btrfs_ioctl_vol_args)
 #define BTRFS_IOC_FORGET_DEV _IOW(BTRFS_IOCTL_MAGIC, 5, \
                                   struct btrfs_ioctl_vol_args)
```

I find it extremely fishy that `BTRFS_IOC_SCAN_DEV` is identical to `BTRFS_IOC_FORGET_DEV`. Deleting a device on scanning would do it.

Proposed fix upstream as https://github.com/kdave/btrfs-progs/pull/706.

**Comment by trofi** [comment]
The upstream fix was released as `6.6.1`. Trying the update again: https://github.com/NixOS/nixpkgs/pull/265745. Please give it a try.
